### ISSUE TITLE
docker run error 'exec format error' post upgrade on Apple silicon M2

### ISSUE THREAD

**Issue by habasse-traore** [docker/for-mac#7168]
### Description

I'm getting an error running 2 images post upgrade to the latest version 4.27.1 of docker-desktop for mac
The container fails with "exec format error" 
Also docker-desktop fails when trying to push an image to docker hub and I have to reset to factory settings before getting it to work again.

### Reproduce

Platform: Macbook Pro M2 
OS: Sonoma 14.2.1
Docker Desktop: 4.27.1
cmd:
1. docker pull gcr.io/k8s-minikube/minikube-ingress-dns:0.0.2
2. docker run gcr.io/k8s-minikube/minikube-ingress-dns:0.0.2
Output: "exec /usr/local/bin/docker-entrypoint.sh: exec format error"
![image](https://github.com/docker/for-mac/assets/144721927/9f9a48e0-f908-4e8a-8eb2-1d28c3de2adb)

1. docker pull docker.io/upmcenterprises/registry-creds:1.10
2. docker run docker.io/upmcenterprises/registry-creds:1.10
Output: "exec /registry-creds: exec format error" 
![image](https://github.com/docker/for-mac/assets/144721927/56811b3b-6f8d-43e6-b28a-6db2fce599b8)

### Expected behavior

docker run should run the container. The same images run on docker-desktop  4.26.1

### docker version

```bash
Client:
 Cloud integration: v1.0.35+desktop.10
 Version:           25.0.2
 API version:       1.44
 Go version:        go1.21.6
 Git commit:        29cf629
 Built:             Thu Feb  1 00:18:45 2024
 OS/Arch:           darwin/arm64
 Context:           desktop-linux

Server: Docker Desktop 4.27.1 (136059)
 Engine:
  Version:          25.0.2
  API version:      1.44 (minimum version 1.24)
  Go version:       go1.21.6
  Git commit:       fce6e0c
  Built:            Thu Feb  1 00:23:21 2024
  OS/Arch:          linux/arm64
  Experimental:     false
 containerd:
  Version:          1.6.28
  GitCommit:        ae07eda36dd25f8a1b98dfbf587313b99c0190bb
 runc:
  Version:          1.1.12
  GitCommit:        v1.1.12-0-g51d5e94
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```bash
Client:
 Version:    25.0.2
 Context:    desktop-linux
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.12.1-desktop.4
    Path:     /Users/habasse.traore/.docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.24.3-desktop.1
    Path:     /Users/habasse.traore/.docker/cli-plugins/docker-compose
  debug: Get a shell into any image or container. (Docker Inc.)
    Version:  0.0.22
    Path:     /Users/habasse.traore/.docker/cli-plugins/docker-debug
  dev: Docker Dev Environments (Docker Inc.)
    Version:  v0.1.0
    Path:     /Users/habasse.traore/.docker/cli-plugins/docker-dev
  extension: Manages Docker extensions (Docker Inc.)
    Version:  v0.2.21
    Path:     /Users/habasse.traore/.docker/cli-plugins/docker-extension
  feedback: Provide feedback, right in your terminal! (Docker Inc.)
    Version:  v1.0.4
    Path:     /Users/habasse.traore/.docker/cli-plugins/docker-feedback
  init: Creates Docker-related starter files for your project (Docker Inc.)
    Version:  v1.0.0
    Path:     /Users/habasse.traore/.docker/cli-plugins/docker-init
  sbom: View the packaged-based Software Bill Of Materials (SBOM) for an image (Anchore Inc.)
    Version:  0.6.0
    Path:     /Users/habasse.traore/.docker/cli-plugins/docker-sbom
  scout: Docker Scout (Docker Inc.)
    Version:  v1.3.0
    Path:     /Users/habasse.traore/.docker/cli-plugins/docker-scout
WARNING: Plugin "/Users/habasse.traore/.docker/cli-plugins/docker-scan" is not valid: failed to fetch metadata: fork/exec /Users/habasse.traore/.docker/cli-plugins/docker-scan: no such file or directory

Server:
 Containers: 0
  Running: 0
  Paused: 0
  Stopped: 0
 Images: 2
 Server Version: 25.0.2
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: ae07eda36dd25f8a1b98dfbf587313b99c0190bb
 runc version: v1.1.12-0-g51d5e94
 init version: de40ad0
 Security Options:
  seccomp
   Profile: unconfined
  cgroupns
 Kernel Version: 6.6.12-linuxkit
 Operating System: Docker Desktop
 OSType: linux
 Architecture: aarch64
 CPUs: 8
 Total Memory: 23.44GiB
 Name: docker-desktop
 ID: a1ec940b-3186-488e-80d1-f6879134a154
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 HTTP Proxy: http.docker.internal:3128
 HTTPS Proxy: http.docker.internal:3128
 No Proxy: hubproxy.docker.internal
 Experimental: false
 Insecure Registries:
  hubproxy.docker.internal:5555
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: daemon is not using the default seccomp profile
```

### Diagnostics ID

EFBFC4C6-AD42-4470-8971-A009AF56D571/20240204153215

### Additional Info

I'm spinning these containers has part of a local minikube dev environment 
the 2 containers images falling are required to enable ingress-dns and registry-creds service for the cluster 
![image](https://github.com/docker/for-mac/assets/144721927/06c0e8bd-b4c9-4e59-aaed-618c6b4527b1)

**Comment by geek-kb** [comment]
I have the same issue.
It started today after updating Docker desktop to the latest version on MacOS M3 Pro Sonoma 14.1 .
Please assist.

**Comment by kryszce** [comment]
I have exactly the same problem after upgrading to the latest docker version 
Version 4.27.1 (136059)
OS Sonoma 14.3

Could someone please advise on how to get this working?

**Comment by fozcode** [comment]
Same with me after upgrade to 4.27:

```
[event: pod futureon/futureon-postgresql-0] Pulling image "docker.io/bitnami/postgresql:15.1.0-debian-11-r12"
[event: pod futureon/futureon-postgresql-0] Successfully pulled image "docker.io/bitnami/postgresql:15.1.0-debian-11-r12" in 12.162s (26.656s including waiting)
exec /opt/bitnami/scripts/postgresql/entrypoint.sh: exec format error
```

Downgrading to 4.26 has fixed it and the image runs again. Download link: https://docs.docker.com/desktop/release-notes/#4261

**Comment by rafaelavd** [comment]
Any update on this? Facing the same issue...

**Comment by fozcode** [comment]
In my case the image is only available for amd64 [link](https://hub.docker.com/layers/bitnami/postgresql/15.1.0-debian-11-r12/images/sha256-ad9b2e94cc1b3fa61b28fa221417c72f3e1ad53cddbd095513ced25a43391487)
```
OS/ARCH
linux/amd64
```
I have Rosetta emulation disabled. So presumably 4.26 was performing some kind of amd64 emulation and 4.27 is not. I'm going to update my image to a newer version that provides both amd64 and arm64.

**Comment by rafaelavd** [comment]
anyone can point out what was the breaking change that caused this?
Thanks

**Comment by mdkent** [comment]
We've got multiple people hitting this here.

Platform: Apple M2 Max
OS: 14.3 (23D56)
Docker Desktop 4.27.1 (136059)

Pretty easy to reproduce with Rosetta enabled:
```
mk-mbp3 3.3.0 ~ docker run --platform linux/amd64 hello-world
Unable to find image 'hello-world:latest' locally
latest: Pulling from library/hello-world
c1ec31eb5944: Pull complete
Digest: sha256:4bd78111b6914a99dbc560e6a20eab57ff6655aea4a80c50b0c5491968cbc2e6
Status: Downloaded newer image for hello-world:latest
exec /hello: exec format error
```
with Rosetta disabled hello-world runs, but I'm getting weird crashes like
```
       qemu: uncaught target signal 11 (Segmentation fault) - core dumped
       Segmentation fault
```
in other apps.

Downgrading to 4.26.1 fixed it as well.

**Comment by adityasamant25** [comment]
I can confirm this as well.
Upgrading to 4.27.1 has broken x86 image emulation on Mac.

Downgrading to 4.26.1 works, so I'll be staying on this version until the bug is fixed on 4.27.x

**Comment by dgageot** [comment]
Hi everyone, I can't reproduce on 4.27.2 that was just released but I think it might be a transient error. Could you try this new version and tell me if it fixes your issue?

**Comment by mmarchini** [comment]
@dgageot still happening consistently on `4.27.2` on macOS `14.3.1`. I can use `docker run --platform linux/amd64 ubuntu:jammy uname -a` to reproduce the error. It also doesn't look device specific since I've had other engineers reporting the same issue today. Could it be at all related to https://github.com/docker/for-mac/issues/7172? Both are related to architecture emulation and both started happening with `4.27.0`.

**Comment by jglogan** [comment]
If you're using any kind of kubernetes node image in your linuxkit VM, it could be because the new node images have some systemd fun that messes up the binfmt handlers.  Since those aren't namespaces, they'll mess up emulation handling for all your containers.

See:
* https://github.com/kubernetes-sigs/kind/issues/3510
* https://github.com/kubernetes/minikube/issues/17700

**Comment by dgageot** [comment]
Thanks a lot @jglogan, that makes a lot of sense. I couldn't figure out how people's binfmt could be messed up that much!

Everyone:

Those of you with the `qemu: uncaught target signal 11 (Segmentation fault) - core dumped
       Segmentation fault` error, you should follow issue https://github.com/docker/for-mac/issues/7172. tl;dr the qemu update we shipped with 4.27 seems to have an issue.

Those of you with the `exec format error`, please can you look at @jglogan's [comment](https://github.com/docker/for-mac/issues/7168#issuecomment-1935284120) and tell me if that applies to you?

**Comment by fozcode** [comment]
> Those of you with the exec format error, please can you look at @jglogan's and tell me if that applies to you?

My use case is with minikube so I think the answer is yes. Thanks for figuring this out!

**Comment by jglogan** [comment]
@fozcode No problem!  TBH I didn't figure it out, a colleague did.  Just passing along what they found!

**Comment by livvius** [comment]
> Upgrading to 4.27.2 seems to have revolved the issue for me Platform: Apple M2 Max OS: Sonoma 14.2.1 Docker Desktop 4.27.2 (137060) Docker-Desktop General Config ![image](https://private-user-images.githubusercontent.com/144721927/303710641-f72354c4-b93a-4035-90d3-b437d5d6e932.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDc1MDQ2MTYsIm5iZiI6MTcwNzUwNDMxNiwicGF0aCI6Ii8xNDQ3MjE5MjcvMzAzNzEwNjQxLWY3MjM1NGM0LWI5M2EtNDAzNS05MGQzLWI0MzdkNWQ2ZTkzMi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjA5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIwOVQxODQ1MTZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02MTUyMGFjODZjYWE5MzA1YzA4ODA1NDYzNDI4OThiYWYyZDcxMzY0ZDQ2OGZkNTFiMjUyNjdiOGQxMzZhN2I4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.K1gxQWUJbDkGSQay-hI13yPdI19YSIS7RTa46D0M15s)
> 
> My use case is with minikube and followed @jglogan comments running `minikube start --base-image gcr.io/k8s-minikube/kicbase-builds:v0.0.42-1703092832-17830 --driver docker` minikube starts now without any issues. The previously failing container images ingress-dns and registry-creds service are now running without `exec format` error ![image](https://private-user-images.githubusercontent.com/144721927/303708331-785adf40-a64a-425f-9f94-d2350d4fc27b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDc1MDQ2MTYsIm5iZiI6MTcwNzUwNDMxNiwicGF0aCI6Ii8xNDQ3MjE5MjcvMzAzNzA4MzMxLTc4NWFkZjQwLWE2NGEtNDI1Zi05Zjk0LWQyMzUwZDRmYzI3Yi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjA5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIwOVQxODQ1MTZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wYTI2YmYzYzZkZjAwZmI5MTU3YzA5NGI3YzJmNWU5ODNlZDYzNzA5YjhmNmRiMDJjZjgzMmE0MWMzMjQ1MDBjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.Bq1GAYS7JyjAOJgt-fxhqwW7NrlEXtZJwqml0zNNy4E)
> 
> Ran an additional validation by running `docker run --platform linux/amd64 hello-world` and that also seems to run fine ![image](https://private-user-images.githubusercontent.com/144721927/303709744-2c3f5db3-4554-4df4-bbbf-2e7582dbba24.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDc1MDQ2MTYsIm5iZiI6MTcwNzUwNDMxNiwicGF0aCI6Ii8xNDQ3MjE5MjcvMzAzNzA5NzQ0LTJjM2Y1ZGIzLTQ1NTQtNGRmNC1iYmJmLTJlNzU4MmRiYmEyNC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjA5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIwOVQxODQ1MTZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02YmFmYjk5YmFhZWUyYjkwMDQzMjg5ZDk4Zjk1NTZmN2JiYjM1NTg3YTY5NmQzMGM5NWQwNDIyMDg5NWFjZmEzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.olIavPXuCGClTZThZHToBhinTx1l5YOtAxRUfqWZ_Cs)

Ingress-dns seems working fine now, but I seems not completly fixed I'm using Percona MongoDB Operator and has the same issue

**Comment by eko** [comment]
I have the same issue here on macOS Sonoma 14.3.1 and Docker Desktop 4.27.2 (Rosetta emulation enabled):

```bash
$ docker run --platform linux/amd64 hello-world

Unable to find image 'hello-world:latest' locally
latest: Pulling from library/hello-world
c1ec31eb5944: Pull complete
Digest: sha256:4bd78111b6914a99dbc560e6a20eab57ff6655aea4a80c50b0c5491968cbc2e6
Status: Downloaded newer image for hello-world:latest
exec /hello: exec format error
```

**Comment by jgmartinez** [comment]
I'm also facing the same issue. Mac M3 Max, MacOS Sonoma 14.3.1, Docker Desktop 4.27.2 with Rosetta enabled.

I fully reinstalled Docker Desktop and it worked for a while, but the issue came back.

```bash
~ docker run --platform linux/amd64 -it ubuntu:latest bash
exec /usr/bin/bash: exec format error
```

**Comment by dgageot** [comment]
I'm going to close this issue. It's not a Docker Desktop issue per se. It's caused by minikube/kind messing up with Docker Desktop's qemu/rosetta configuration.

This [comment](https://github.com/docker/for-mac/issues/7168#issuecomment-1936964854) gives a solution. Basically, Minikube needs to be started with a newer image that was fixed to not mess with Docker Desktop.

**Comment by biscout42** [comment]
Same issue here, Mac M2 Pro, MacOS Sonoma 14.3.1, Docker Desktop 4.27.2 with Rosetta enabled.

Not related to `minikube`.  Following command fails on DD `4.27.2` 

```
~ docker run --platform linux/amd64 -it ubuntu:latest bash
exec /usr/bin/bash: exec format error
```

And it works with DD `4.26.1`

**Comment by jlmoya** [comment]
Same as @biscout42 , After upgrading to 4.27.x all AMD images fail to start with "exec format error" on Mac M2 with Sonoma 14.3.1.  Tested with 4.27.0, 4.27.1 and 4.27.2 all fail with the same "exec format error" message.  Falling back to 4.26.1 works flawlessly.

**Comment by biscout42** [comment]
@dgageot , would you suggest to create a new issue or re-open this one? As I see the topic was also about docker w/o minikube (only image name)
```
docker pull gcr.io/k8s-minikube/minikube-ingress-dns:0.0.2
```

**Comment by vietvudanh** [comment]
I don't think this problem caused by `minikube`, I have deleted, using new image for `minikube`, even remove `minikube` all together but still got the error. Others stated that they do not use `minikube` too.

**Comment by dgageot** [comment]
@biscout42 @jlmoya @vietvudanh Please open another issue if you really think it's not related to minikube/k8s/kind.
I have to admit I have doubts since at least two of you have either installed `minikube` or ran minikube images.
On the new issue, please share your diagnostics so that I can try to understand. Thanks!

**Comment by biscout42** [comment]
Double checked again, and imho @dgageot , you have a point. If I remove minikube via
```
brew remove minikube
```

and then clean and restart DD
```
docker system prune -a
```

I can run following command w/o errors on `Docker Desktop 4.27.2 (137060)` 

```
docker run --platform linux/amd64 -it ubuntu:latest bash
```

**Comment by biscout42** [comment]
FYI: https://github.com/kubernetes/minikube/issues/18217

**Comment by vietvudanh** [comment]
Oh ok, my bad. I didn't prune everything after remove minikube. 

I did what @biscout42 suggest in [his comment](https://github.com/docker/for-mac/issues/7168#issuecomment-1953859233) and now it looks fine.

Thank you @dgageot and @biscout42.

**Comment by tillpos-tony** [comment]
fixed and can be run in docker desktop 4.38.0
